### PR TITLE
Lanczos Solver `which=SA,SM,LA,LM` argument

### PR DESCRIPTION
**Pull request by aamijar**

### CODE REVIEW

File: cpp/include/raft/sparse/solver/detail/lanczos.cuh [hunk 1]
```diff
@@ -1769,10 +1772,80 @@ auto lanczos_smallest(
...
+      // Take the first nEigVecs indices (smallest magnitude)
+      raft::copy(selected_indices.data_handle(), indices.data_handle(), nEigVecs, stream);
+    } else if (which == LANCZOS_WHICH::LM) {
+      // Take the last nEigVecs indices (largest magnitude)
+      raft::copy(
+        selected_indices.data_handle(), indices.data_handle() + (ncv - nEigVecs), nEigVecs, stream);
+    }
+
+    // Re-sort these indices by algebraic value to maintain algebraic ordering
+    thrust::sort(thrust::device,
```
Comment (lowener) [inline cpp/include/raft/sparse/solver/detail/lanczos.cuh:1821]: cupy's version is doing only one argsort. Is this one absolutely necessary?

Comment (aamijar) [reply cpp/include/raft/sparse/solver/detail/lanczos.cuh:1821]: Yes, since cupy actually does two argsort as well. One in the _eigsh_solve_ritz function and then one in the eigsh before returning

File: cpp/include/raft/sparse/solver/lanczos_types.hpp [hunk 2]
```diff
@@ -20,6 +20,12 @@
 
 namespace raft::sparse::solver {
 
+// LA: Largest (algebraic) eigenvalues
```
Comment (lowener) [inline cpp/include/raft/sparse/solver/lanczos_types.hpp:23]: Can this be in doxygen format? Right now checking the doxygen-generated doc this doesn't appear.

**Review by lowener — APPROVED** [review]
LGTM

File: cpp/include/raft/sparse/solver/lanczos_types.hpp (hunk 2 above)
Comment (cjnolet) [reply cpp/include/raft/sparse/solver/lanczos_types.hpp:23]: +1

File: cpp/tests/sparse/solver/lanczos.cu [hunk 3]
```diff
@@ -346,6 +348,7 @@ const std::vector<lanczos_inputs<int, float>> inputsf = {
    0,
    0,
    1e-15,
+   raft::sparse::solver::LANCZOS_WHICH::SA,
```
Comment (cjnolet) [inline cpp/tests/sparse/solver/lanczos.cu:351]: We really need more tests here. Coverage is key here (both gtest and Python). We want to make sure we're covering all basic cases as well as potential edge cases. We want to find bugs before users do.

File: cpp/include/raft/sparse/solver/lanczos_types.hpp [hunk 4]
```diff
@@ -20,6 +20,21 @@
...
+  /** @brief LM: Largest (in magnitude) eigenvalues */
+  LM,
+  /** @brief SA: Smallest (algebraic) eigenvalues */
+  SA,
+  /** @brief SM: Smallest (in magnitude) eigenvalues */
+  SM
+};
+
 template <typename ValueTypeT>
 struct lanczos_solver_config {
```
Comment (lowener) [inline cpp/include/raft/sparse/solver/lanczos_types.hpp:39]: Add this to doxygen too please

File: cpp/tests/sparse/solver/lanczos.cu (hunk 3 above)
Comment (aamijar) [reply cpp/tests/sparse/solver/lanczos.cu:656]: More test coverage has been added

File: cpp/tests/sparse/solver/lanczos.cu [hunk 5]
```diff
@@ -439,32 +748,184 @@ const std::vector<lanczos_inputs<int, double>> inputsd = {
...
+INSTANTIATE_TEST_CASE_P(LanczosTests, LanczosTestD_LA, ::testing::ValuesIn(inputsd_LA));
+INSTANTIATE_TEST_CASE_P(LanczosTests, LanczosTestD_LM, ::testing::ValuesIn(inputsd_LM));
+INSTANTIATE_TEST_CASE_P(LanczosTests, LanczosTestD_SA, ::testing::ValuesIn(inputsd_SA));
+
+INSTANTIATE_TEST_CASE_P(LanczosTests, LanczosTestF_SM, ::testing::ValuesIn(inputsf_SM));
+INSTANTIATE_TEST_CASE_P(LanczosTests, LanczosTestF_LA, ::testing::ValuesIn(inputsf_LA));
+INSTANTIATE_TEST_CASE_P(LanczosTests, LanczosTestF_LM, ::testing::ValuesIn(inputsf_LM));
+INSTANTIATE_TEST_CASE_P(LanczosTests, LanczosTestF_SA, ::testing::ValuesIn(inputsf_SA));
+
 }  // namespace raft::sparse
```
Comment (aamijar) [inline cpp/tests/sparse/solver/lanczos.cu:931]: change namespace to be raft::sparse::solver

File: python/pylibraft/pylibraft/tests/test_sparse.py [hunk 6]
```diff
@@ -98,13 +98,14 @@ def _test_eigsh(self, a, k, xp, sp):
     @pytest.mark.parametrize("format", ["csr"])  # , 'csc', 'coo'])
     @pytest.mark.parametrize("k", [3, 6, 12])
     @pytest.mark.parametrize("dtype", ["f", "d"])
-    def test_sparse(self, format, k, dtype, xp=cupy, sp=sparse):
+    @pytest.mark.parametrize("which", ["LA", "LM", "SA"])
```
Comment (cjnolet) [inline python/pylibraft/pylibraft/tests/test_sparse.py:101]: Is SM being tested anywhere in the Python layer?

Comment (aamijar) [reply python/pylibraft/pylibraft/tests/test_sparse.py]: Pytest updated

**Comment by aamijar** [comment]
Hi @cjnolet,  I'm running into some issues with tests not passing when using solver with `SM`. Spectral embedding only needs `LA`, so in the interest of getting that in, can we follow up with `SM` in a separate PR?

**Review by divyegala — APPROVED** [review]
Can you create a follow-up issue for addressing `SM` failures and add it as a sub-issue to the overarching tracker issue?

**Comment by aamijar** [comment]
> Can you create a follow-up issue for addressing `SM` failures and add it as a sub-issue to the overarching tracker issue?

done!
